crates/core/src/sql/kind.rs

```diff
@@ -242,6 +257,137 @@ impl Kind {
 	}
 }
 
+/// Trait for retrieving the `kind` equivalent of a rust type.
+///
+/// Returns the most general kind for a type.
+/// For example Number could be either number or float or int or decimal but the most general is
+/// number.
+///
+/// This trait is only implemented for types which can only be retrieve from
+pub trait HasKind {
+	fn kind() -> Kind;
+}
+
+impl<T: HasKind> HasKind for Option<T> {
+	fn kind() -> Kind {
+		let kind = T::kind();
+		if matches!(kind, Kind::Option(_)) {
+			kind
+		} else {
+			Kind::Option(Box::new(kind))
+		}
+	}
+}
+
+impl<T: HasKind> HasKind for Vec<T> {
+	fn kind() -> Kind {
+		let kind = T::kind();
+		Kind::Array(Box::new(kind), None)
+	}
+}
+
+impl HasKind for Array {
+	fn kind() -> Kind {
+		Kind::Array(Box::new(Kind::Any), None)
+	}
+}
+
+impl<T: HasKind, const SIZE: usize> HasKind for [T; SIZE] {
+	fn kind() -> Kind {
+		let kind = T::kind();
+		Kind::Array(Box::new(kind), Some(SIZE as u64))
+	}
+}
+
+impl HasKind for Thing {
+	fn kind() -> Kind {
+		Kind::Record(Vec::new())
+	}
+}
+
+impl HasKind for Geometry {
+	fn kind() -> Kind {
+		Kind::Geometry(Vec::new())
+	}
+}
+
+impl HasKind for Closure {
+	fn kind() -> Kind {
+		// The inner values of function are currently completely unused.
+		Kind::Function(None, None)
+	}
+}
+
+impl HasKind for Regex {
+	fn kind() -> Kind {
+		Kind::Regex
+	}
+}
+
+impl HasKind for File {
+	fn kind() -> Kind {
+		Kind::File(Vec::new())
+	}
+}
+
+macro_rules! impl_basic_has_kind{
+	($($name:ident => $kind:ident),*$(,)?) => {
+		$(
+			impl HasKind for $name{
+				fn kind() -> Kind{
+					Kind::$kind
+				}
+			}
+		)*
+	}
+}
+
+impl_basic_has_kind! {
+	bool => Bool,
+
+	// Only numbers which fully convert to an int
+	i64 => Int,
+	i32 => Int,
+	i16 => Int,
+	i8 => Int,
+	u32 => Int,
+	u16 => Int,
+	u8 => Int,
+	f64 => Float,
```

Comment:
Should `f32` be in here?
```suggestion
	f64 => Float,
    f32 => Float,
```

Reply:
I didn't want to implement it for f32 as there is no direct analogous type in SurrealQL for f32. All floats are f64.

Reply:
That makes sense to me, but then based on that logic why are `u8`, `u16`, `u32`, `i8`, `i16`, and `i32` implemented for `Int` if the only representation for `Int` is `i64`?

Reply:
I probably decided that only the direct types should have been implemented after I added the integers. Will remove them.